design/design-generic-references.md

```diff
@@ -0,0 +1,319 @@
+# Generic Resource References
+
+* Owner: Predrag Knezevic (@pedjak)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane providers provision and monitor external resources in third party
+cloud providers, following the declared specification in the corresponding managed
+resources. Very often, creation of these objects cannot happen in parallel, i.e. some
+fields can be set only after a given dependency is created.
+
+[Cross resource referencing] solves this problem for majority of cases by
+following [Kubernetes API object reference] convention:
+
+```yaml
+spec:
+  fooRef:
+    name: bar
+```
+
+Field `spec.fooRef` contains the name of the resource where the value for `spec.foo` is to
+be found. The kind of referred resource and the field containing the value used later for
+setting `spec.foo` is determined during compile time.
+
+Although the used convention covers the majority of use cases, it lacks the
+following properties:
+
+* The referred object kind and the source field cannot be changed at 
+  runtime for a given object
+* The referencer field may not always target a single referred kind. For
+  example, a URL field in Route53 could get populated by a URL of an S3 bucket
+  or access endpoint of an RDS Instance
+* The referencer field may not always target a single field in the referred
+  object
+* The value that is about to be assigned to the field require some sort of
+  transformation
+
+Some of the above issues can be solved by defining a composition. However, there
+are cases where one would like to use and refer to a resource not bound by a 
+composite instance.
+
+__**NOTE**__: [a prior design exists on this topic](https://github.com/crossplane/crossplane/pull/2385).
+
+## Goals
+
+The generic cross-reference should be able to:
+
+* be configurable at runtime
+* use the existing [Kubernetes API object reference] convention 
+  as much as possible
+* provide alternative convention in cases where [Kubernetes API object reference]
+  convention is not suitable
+* be implemented without or requiring only light changes in RBAC
+* support referring multiple objects (namespaced and cluster-scoped) and their
+  fields
+* support value transformation before assignment
+
+## Proposal
+
+We would like to introduce a new cluster-scoped `Referable` (alternative name
+could be `InjectableValue`) type whose instances declare values that can be
+assigned to an object field:
```

Comment:
alternative: name it by what these things are namely `FieldReferences`, `ObjectReferences`. Naming something by how it can be used is a little strange.

Reply:
+1